### PR TITLE
Turn into real ES Node module

### LINKED ISSUE #48: Add true ES module
Add similar project structure to what https://github.com/fregante/github-url-detection does so that people can directly import from ES module projects.

### PR DESCRIPTION
**Pull request by notlmn**
Closes #48

Shamelessly copied code over from https://github.com/fregante/github-url-detection because the state of ES module is a friggin mess right now.

Haven't tested this yet though, but it should work fine.

Sample `npm pack`:
```
npm notice 📦  webext-options-sync@1.2.3
npm notice === Tarball Contents === 
npm notice 1.1kB  license         
npm notice 26.7kB cjs/index.js    
npm notice 26.7kB esm/index.js    
npm notice 22B    esm/package.json
npm notice 1.6kB  package.json    
npm notice 9.1kB  readme.md       
npm notice 3.2kB  cjs/index.d.ts  
npm notice 3.2kB  esm/index.d.ts  
npm notice === Tarball Details === 
npm notice name:          webext-options-sync                     
npm notice version:       1.2.3                                   
npm notice filename:      webext-options-sync-1.2.3.tgz           
npm notice package size:  11.0 kB                                 
npm notice unpacked size: 71.6 kB                                 
npm notice shasum:        2c6cccbf80f18bd5cbc7ac2034d0d8dad3d128b1
npm notice integrity:     sha512-Qlw7e65RTyBUk[...]SA4sWujTi8vdg==
npm notice total files:   8                                       
npm notice 
webext-options-sync-1.2.3.tgz
```

### CODE REVIEW

**Comment by fregante** [comment]
Two issues:

- I actually had it before and dropped it (probably because my dependency was not a true ESM either) https://github.com/fregante/webext-options-sync/commit/35eb290b14aa2af3c2883c4edecd2443fd357c7d
- Also `webext-options-sync-per-domain` would also have to be updated to support RGH

If this *can* be a true ESM, then it should ONLY be ESM following https://github.com/fregante/meta/issues/1

**Comment by fregante** [comment]
I double-checked. 2 out of 3 dependencies are not true ESM so I can’t convert this one either. 

The only way to convert would be to change the dependencies or vendor them (😬).

Following https://github.com/fregante/meta/issues/1 also basically mean that almost no changes in this PR are applicable.

**Comment by fregante** [comment]
Well I’m a bit silly. Rollup solves exactly that problem. So this PR only needs to drop the CJS bundle and just make it ESM by default

File: rollup.config.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	output: {
+		format,
+		dir: format
+	},
+	plugins: [
+		commonjs(),
+		typescript({
+			outDir: format
+		}),
+		terser({
```
Comment (fregante) [inline rollup.config.js:16]: We can drop terser probably. The reason why it existed in that module is that it had inline tests that I wanted to drop before publishing.

**Comment by fregante** [comment]
Can you also *un-vendor* this dependency? https://github.com/fregante/webext-options-sync/blob/master/vendor/lz-string.js

File: rollup.config.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+				beautify: true
+			},
+			mangle: false,
+			compress: {
+				join_vars: false, // eslint-disable-line camelcase
+				booleans: false,
+				expression: false,
+				sequences: false
+			}
+		})
```
Comment (fregante) [inline rollup.config.js:29]:
```suggestion
		})
```

File: rollup.config.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import typescript from '@rollup/plugin-typescript';
+import {terser} from 'rollup-plugin-terser';
```
Comment (fregante) [inline rollup.config.js:2]:
```suggestion
```

File: package.json [hunk 4]
```diff
@@ -44,22 +48,27 @@
...
 		"@types/dom-inputevent": "^1.0.5",
 		"@types/lz-string": "^1.3.34",
 		"@types/throttle-debounce": "^2.1.0",
-		"ava": "^2.4.0",
+		"ava": "^3.10.1",
 		"esm": "^3.2.25",
-		"sinon": "^9.0.1",
+		"npm-run-all": "^4.1.5",
+		"rollup": "^2.23.0",
+		"rollup-plugin-terser": "^6.1.0",
```
Comment (fregante) [inline package.json:67]:
```suggestion
```

File: package.json [hunk 5]
```diff
@@ -44,22 +48,26 @@
...
+		"npm-run-all": "^4.1.5",
+		"rollup": "^2.23.0",
+		"sinon": "^9.0.2",
 		"sinon-chrome": "^3.0.1",
-		"type-fest": "^0.13.0",
-		"typescript": "^3.8.3",
-		"xo": "^0.28.3"
+		"type-fest": "^0.16.0",
+		"typescript": "^3.9.7",
+		"xo": "^0.32.1"
```
Comment (fregante) [inline package.json:71]: It's best to leave any unrelated dependency update out of this PR unless necessary (AVA?).

I/you can update in separate PR

File: rollup.config.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+		format,
+		dir: format
+	},
+	plugins: [
+		commonjs(),
+		typescript({
+			outDir: format
+		})
+	]
+}));
```
Comment (fregante) [inline rollup.config.js:16]: I think? Maybe one of the two `'.'` can be dropped

```suggestion
export default {
	input: 'index.ts',
	output: {
		format: 'esm',
		dir: '.'
	},
	plugins: [
		commonjs(),
		typescript({
			outDir: '.'
		})
	]
};
```

Comment (notlmn) [reply rollup.config.js:16]: We need both, one for `index.js` and one for `index.d.ts`

**Comment by fregante** [comment]
## Base: 39kB

```js
npm notice 
npm notice 📦  webext-options-sync@1.2.3
npm notice === Tarball Contents === 
npm notice 1.1kB  license     
npm notice 39.6kB index.js    
npm notice 1.6kB  package.json
npm notice 9.1kB  readme.md   
npm notice 3.2kB  index.d.ts  
npm notice === Tarball Details === 
npm notice name:          webext-options-sync                     
npm notice version:       1.2.3                                   
npm notice filename:      webext-options-sync-1.2.3.tgz           
npm notice package size:  14.1 kB                                 
npm notice unpacked size: 54.5 kB                                 
npm notice shasum:        c03758198e70dc8be155d14b5780e1993a379c13
npm notice integrity:     sha512-4PhGKSeWRUcJW[...]yxIOPJEtYbj8Q==
npm notice total files:   5                                       
npm notice 
```

## + vendored `lz-string`: 34kB

```js
npm notice 
npm notice 📦  webext-options-sync@1.2.3
npm notice === Tarball Contents === 
npm notice 1.1kB  license     
npm notice 34.4kB index.js    
npm notice 1.6kB  package.json
npm notice 9.1kB  readme.md   
npm notice 3.2kB  index.d.ts  
npm notice === Tarball Details === 
npm notice name:          webext-options-sync                     
npm notice version:       1.2.3                                   
npm notice filename:      webext-options-sync-1.2.3.tgz           
npm notice package size:  13.3 kB                                 
npm notice unpacked size: 49.4 kB                                 
npm notice shasum:        056abfdba23be5ac7fd9ac799d3c13e901c29565
npm notice integrity:     sha512-m3nCT3O6YjRVK[...]FwwgKXWKIgHJw==
npm notice total files:   5                                       
npm notice 
```

## + minifier: 28kB

```js
npm notice 
npm notice 📦  webext-options-sync@1.2.3
npm notice === Tarball Contents === 
npm notice 1.1kB  license     
npm notice 28.0kB index.js    
npm notice 1.6kB  package.json
npm notice 9.1kB  readme.md   
npm notice 3.2kB  index.d.ts  
npm notice === Tarball Details === 
npm notice name:          webext-options-sync                     
npm notice version:       1.2.3                                   
npm notice filename:      webext-options-sync-1.2.3.tgz           
npm notice package size:  10.8 kB                                 
npm notice unpacked size: 43.0 kB                                 
npm notice shasum:        01fd00cd4675bd841d39fea64600e89c1629f784
npm notice integrity:     sha512-qgz5TCt8SLR+C[...]yIQm4gjSFAQsg==
npm notice total files:   5                                       
npm notice 
```

It's worth keeping both things as they are. Perhaps one day I'll be able to drop the serializer, but it's not easy... https://github.com/fregante/webext-options-sync/pull/28#issuecomment-514959444

**Comment by fregante** [comment]
You can test it in your PR with:

```js
npm install webext-options-sync@next
```

**Comment by fregante** [comment]
Note: WOS is already an ES Module, just not a Node ES Module. This PR is missing `type:module, main:index.js`

I'm gonna run some tests and add it

**Comment by fregante** [comment]
https://github.com/fregante/webext-options-sync-per-domain/pull/1 also seems to be imported/tested correctly in RG under Node 14

**Comment by fregante** [comment]
Well, everything seems to work! Including #52 

The tests still aren't compatible with Node 14, but that can be separate. Related: https://github.com/vadimdemedes/dom-chef/pull/68 (but this package won't have the same problem)

[webext-options-sync-per-domain](https://github.com/fregante/webext-options-sync-per-domain/pull/1) would also need the same thing due to its `mem` dependency and can't be `type:module` until then (even if I tested it and it works in RG)

**Comment by fregante** [comment]
This will be released as a breaking release. _If anyone can show just cause why this module cannot lawfully be published as `breaking`, let them speak now or forever hold their peace._

**Comment by notlmn** [comment]
>  _If anyone can show just cause why this module cannot lawfully be published as `breaking`, let them speak now or forever hold their peace._

I blame Node ESM again!

**Pull request merged**